File: src/ee/ads/MultiRewardedAd.cpp
//
//  RewardedVideoList.cpp
//  ee_x
//
//  Created by Zinge on 10/11/17.
//
//

#include "ee/ads/MultiRewardedAd.hpp"

#include <ee/core/ObserverHandle.hpp>
#include <ee/coroutine/NoAwait.hpp>
#include <ee/coroutine/Task.hpp>

namespace ee {
namespace ads {
using Self = MultiRewardedAd;

Self::MultiRewardedAd() {
    handle_ = std::make_unique<ObserverHandle>();
}

Self::~MultiRewardedAd() {
    handle_->clear();
}

Self& Self::addItem(const std::shared_ptr<IRewardedAd>& item) {
    items_.push_back(item);
    (*handle_).bind(*item).addObserver({
        .onLoaded =
            [this] {
                // Propagation.
                dispatchEvent([](auto&& observer) {
                    if (observer.onLoaded) {
                        observer.onLoaded();
                    }
                });
            },
        .onClicked =
            [this] {
                // Propagation.
                dispatchEvent([](auto&& observer) {
                    if (observer.onClicked) {
                        observer.onClicked();
                    }
                });
            },
    });
    return *this;
}

void Self::destroy() {
    for (auto&& item : items_) {
        item->destroy();
    }
    items_.clear();
    handle_->clear();
}

bool Self::isLoaded() const {
    for (auto&& item : items_) {
        if (item->isLoaded()) {
            return true;
        }
    }
    return false;
}

Task<bool> Self::load() {
    bool result = false;
    for (auto&& item : items_) {
        if (item->isLoaded()) {
            continue;
        }
        if (co_await item->load()) {
            result = true;
        }
    }
    co_return result;
}

Task<IRewardedAdResult> Self::show() {
    auto result = IRewardedAdResult::Failed;
    for (auto&& item : items_) {
        if (result == IRewardedAdResult::Failed) {
            // Only process if there isn't any successfully displayed ad.
            result = co_await item->show();
        }
        if (not item->isLoaded()) {
            noAwait([item]() -> Task<> { //
                // Load in background.
                co_await item->load();
            });
        }
    }
    co_return result;
}
} // namespace ads
} // namespace ee

File: unity/Runtime/Ads/MultiRewardedAd.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace EE {
    public class MultiRewardedAd : ObserverManager<IRewardedAdObserver>, IRewardedAd {
        private readonly List<IRewardedAd> _items;
        private readonly ObserverHandle _handle;

        public MultiRewardedAd() {
            _items = new List<IRewardedAd>();
            _handle = new ObserverHandle();
        }

        public MultiRewardedAd AddItem(IRewardedAd item) {
            _items.Add(item);
            _handle.Bind(item).AddObserver(new IRewardedAdObserver {
                OnLoaded = () => DispatchEvent(observer => observer.OnLoaded?.Invoke()),
                OnClicked = () => DispatchEvent(observer => observer.OnClicked?.Invoke())
            });
            return this;
        }

        public void Destroy() {
            foreach (var item in _items) {
                item.Destroy();
            }
            _items.Clear();
            _handle.Clear();
        }

        public bool IsLoaded => _items.Any(item => item.IsLoaded);

        public async Task<bool> Load() {
            var result = false;
            foreach (var item in _items) {
                if (item.IsLoaded) {
                    continue;
                }
                if (await item.Load()) {
                    result = true;
                }
            }
            return result;
        }

        public async Task<IRewardedAdResult> Show() {
            var result = IRewardedAdResult.Failed;
            foreach (var item in _items) {
                if (result == IRewardedAdResult.Failed) {
                    // Only process if there isn't any successfully displayed ad.
                    result = await item.Show();
                }
                if (!item.IsLoaded) {
                    // Load in background.
                    _ = item.Load();
                }
            }
            return result;
        }
    }
}